README.md
# Variable-Solutions #

The Variable-Solutions plugin adds global variables to the WordPress content management system.

## Description ##

The Variable-Solutions plugin extends WordPress enabling the creation and management of variables that can be referenced as short codes in WordPress content and/or as defined variables in the supporting PHP files. It is a thin plugin that loads only one file when running in end user mode. The Variable-Solutions plugin is developed and maintained by <a href="https://www.usi2solve.com">Universal Solutions</a>.

**What are Variables** - Say you have some information that changes frequently and appears on multiple pages.
It could be the date of your next big event or even the product of the month. 
Keeping track of where this information appears on your site and making consistent updates can be a labor intensive and error prone process.
The Variable-Solutions plugin uses shortcodes to give each piece of information a unique "*variable*" name. 
These unique variable names, along with the information the name represents, are easily managed in the admin back end.

**What are Shortcodes** - Shortcodes are special tags used as shortcuts to easily insert bits of functionality into your content. 
You can recognize shortcodes by square brackets that surround a simple word or a phrase.

**Show Me an Example** - If your site features a product of the month, you can create the variable ` 'product_of_the_month' ` and set it to ` "Our New Gadget" `, 
then you can include this variable anywhere in your content with the following shortcode:

```
[variable item="product_of_the_month"]
```

and your visitors will see ` "Our New Gadget" ` wherever the above shortcode is used. 
Next month when your product changes, you can make one edit and set the ` 'product_of_the_month' ` variable to ` "Our Newest Gadget" ` 
and your visitors will see the updated content everywhere the shortcode is used.

## Installation ##
The Variable-Solutions plugin follows the standard WordPress <a href="https://codex.wordpress.org/Managing_Plugins#Manual_Plugin_Installation">manual plugin installation</a> procedure:
1. Install the Settings-Solutions plugin.
1. Clone or Download the Variable-Solutions archive to your computer.
1. Extract the archive contents to your local file system.
1. Rename the extracted folder to ` usi-variable-solutions ` if not already done so during the extraction.
1. Upload the ` .php ` files from your ` usi-variable-solutions ` folder to the ` wp-content/plugins ` folder in your target WordPress installation.
1. Activate the plugin via the WordPress *Plugins* menu located on the left side bar.

## Implementation ##
The Variable-Solutions plugin stores the variables you create in the WordPress database and writes them to the *variables.php* file. 
This file contains PHP define statements that can be referenced by code in your system.
Using the example from above, the *variables.php* file would contain the following line:

```php
define('prefix_PRODUCT_OF_THE_MONTH', 'Our New Gadget');
```
Any custom theme or plugin code in your WordPress installation can reference the `prefix_PRODUCT_OF_THE_MONTH ` variable and get the value `'Our New Gadget'`.
The ` prefix_ ` string defaults to the WordPress database prefix.

You can control the location of the *variables.php* file and other configuration options via the *Variable-Solutions* page under the WordPress *Settings* 
menu located on the left side bar.

## Settings ##
The Variable-Solutions settings page contains three tabs: Preferences, Publish and Capabilities.

**Preferences** - allows you to change the default options set during plugin installation.

**Publish** - any changes you make to you variables must be published before the changes appear to the world.

**Capabilities** - allows you to set the role capabilites system-wide or for a specific user on a user-by-user basis. 

## Usage ##
Click on the WordPress *Variables* menu located on the left side bar to see the list of variables that are in your WordPress installation.
Click the ` Add New ` button to add a new variable or hover over a variable name and click the ` Edit ` link to edit an existing variable.

The edit variable page shows the variable category and variable name for the selected variable along with some other parameters. 
The Variable-Solutions plugin comes with the built-in ` general ` category which is used as the default category if no category is given. 
There is no harm in specifying the ` general ` category, in which case the shortcode is written as follows:

```
[variable category="general" item="product_of_the_month"]
```

**Note** - If you edit an existing variable and change it's category or variable name, then the old category and variable name pair are repalced with the new combination.
This can cause referencing errors if you don't make the corresponding changes in your content.

Additionally, there is the built-in ` date ` category for inserting dates into your content. 
Enter the date using the 24-hour clock, year first then time format as follows:

```
1970-01-01 00:00:00
```
You can then specify how the date is displayed when WordPress processes the shortcode. For example:

```
[variable category="date" item="unix_epoch" format="F js, Y"]
```

displays ` January 1st, 1970 `, which is the format specified by the ` F js, Y ` string. 
You can find the format string parameter reference here: <a href="http://php.net/manual/en/function.date.php">php.net/manual/en/function.date.php</a>.

## Screen Images ##

![Variable List Page](https://user-images.githubusercontent.com/16763256/32197597-40eb1c98-bd9b-11e7-938e-cf46aefd9973.png "Variable List Page")
Screen Image 1 - Variable List Page

<br>

![Add Variable Page](https://user-images.githubusercontent.com/16763256/32197596-40df18b2-bd9b-11e7-8477-3a460bbe7740.png "Add Variable Page")
Screen Image 2 - Add Variable Page

## License ##
> Variable-Solutions is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License 
as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

> Variable-Solutions is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

> You should have received a copy of the GNU General Public License along with Variable-Solutions.  If not, see 
<http://www.gnu.org/licenses/>.

## Donations ##
Donations are accepted at <a href="https://www.usi2solve.com/donate/variable-solutions">www.usi2solve.com/donate</a>. Thank you for your support!

CHANGELOG.md
# Change Log #

Variable-Solutions plugin changes are logged here using <a href="http://semver.org/">Semantic Versioning</a>.

## 2.5.0 (2023-07-04) ##
* Cleaned up code, set all version numbers to same version.

## 2.4.9 (2023-06-08) ##
* Added e-mail variables and obfuscation.

## 2.4.8 (2023-05-19) ##
* Removed GitHub update functionality.

## 2.4.7 (2023-05-18) ##
* Modified so no errors thrown if variables.pohp file is not writen.

## 2.4.6 (2022-07-12) ##
* Fixed menu positioning that was pitching an error.

## 2.4.5 (2022-06-28) ##
* Removed extra underscore character.

## 2.4.4 (2022-02-22) ##
* Set all version numbers to same version.

## 2.4.3 (2021-11-10) ##
* Added usage notes in settings.

## 2.4.2 (2021-11-03) ##
* Removed '@' in include file area so errors show up on admin screen.
* Set all version numbers to same version.

## 2.4.1 (2021-10-30) ##
* Fixed initialization error.

## 2.4.0 (2021-02-24) ##
* Set all version numbers to same version.

## 2.3.2 (2021-02-04) ##
* Added die message if not accessed by WordPress code.

## 2.3.1 (2021-01-07) ##
* Addressed issue with the publish button on the save/edit variable page.

## 2.3.0 (2020-09-14) ##
* Added forced update capabilities to the USI_WordPress_Update class.

## 2.2.0 (2020-06-16) ##
* Resolved permission and capability issues, reworked install, activation and deactivation scheme, set all version numbers to same version.

## 2.1.2 (2020-03-31) ##
* Added use of WordPress-Solutions static class for 'action_admin_head'.

## 2.1.1 (2020-03-22) ##
* Added use of WordPress-Solutions static class for 'column_style'.

## 2.1.0 (2020-02-20) ##
* Updated capability and updates handling, set all version numbers to same version.

## 2.0.5 (2020-02-04) ##
* Improved update handling.

## 2.0.4 (2020-02-03) ##
* Changed code to handle multiple repository sources for updates.

## 2.0.3 (2020-01-07) ##
* Changed format of version number on main plugin file header to facilitate Git Hub update.

## 2.0.2 (2020-01-06) ##
* Changed format of version number.

## 2.0.1 (2020-01-04b) ##
* Added .gitattributes to exclude *.md files from release.

## 2.0.0 (2020-01-04) ##
* Updated license and copyright statements.

## 1.2.1 (2020-01-01) ##
* Added 'Updates' tab to settings page to optionally allow GitHub updates.

## 1.2.0 (2019-12-29) ##
* Added WordPress-Solutions updated for downloading directly from GIT.
* Updated all plugin version numbers to match.

## 1.1.1 (2019-06-12) ##
* Removed usi-settings files from sub-folder under parent plugin and made a stand alone class.
* Updated all plugin version numbers to match.

## 1.0.8 (2018-01-10) ##
* Moved usi-settings files into their own folder and made them a Git submodule.

## 1.0.5 (2018-01-07) ##
* Updated USI_Settings_Admin to change scope of class properties and added options to page_render().

## 1.0.6 (2017-12-14) ##
* Added version list to plugins page.

## 1.0.5 (2017-12-10) ##
* Fixed undefined variable refrence in the USI_Variable_Solutions_Variable::fields_sanitize() method.

## 1.0.4 (2017-12-07) ##
* Fixed the tab reference string in the USI_Variable_Solutions_Settings::fields_sanitize() method.

## 1.0.0 (2017-10-29) ##
* Initial release.



usi-variable-solutions-table.php
<?php // ------------------------------------------------------------------------------------------------------------------------ //

defined('ABSPATH') or die('Accesss not allowed.');

if (!class_exists('WP_List_Table')) { require_once(ABSPATH . 'wp-admin/includes/class-wp-list-table.php'); }

require_once(WP_PLUGIN_DIR . '/usi-wordpress-solutions/usi-wordpress-solutions-static.php');

final class USI_Variable_Solutions_Table extends WP_List_Table {

   const VERSION = '2.5.0 (2023-07-04)';

   private $all_categories = null;
   private $category = null;
   private $options_category = null;
   private $page_hook = null;

   function __construct() {

      $this->options_category = USI_Variable_Solutions::PREFIX . '-options-category';

      add_action('admin_menu', array($this, 'action_admin_menu'));

      if (!empty($_GET['page']) && (USI_Variable_Solutions::VARYLIST == $_GET['page'])) {

         add_action('admin_head', array($this, 'action_admin_head'));

         add_filter('set-screen-option', array($this, 'filter_set_screen_options'), 10, 3);

      }

      $this->category = $this->all_categories = __('All Categories', USI_Variable_Solutions::TEXTDOMAIN);

   } // __construct();

   function action_admin_head() {

      if (USI_Variable_Solutions::VARYLIST != ((isset($_GET['page'])) ? $_GET['page'] : '')) return;

      $columns = array(
         'cb'          => 3, 
         'variable_id' => 7, 
         'variable'    => 15, 
         'category'    => 10, 
         'type'        => 10, 
         'value'       => 15, 
         'notes'       => 15, 
         'owner'       => 10, 
         'updated'     => 15,
      );

      echo USI_WordPress_Solutions_Static::column_style($columns, 'overflow:hidden; text-overflow:ellipsis; white-space:nowrap;');

   } // action_admin_head();

   function action_admin_menu() {

      $capability = USI_WordPress_Solutions_Capabilities::capability_slug(USI_Variable_Solutions::PREFIX, 'view-variables');

      $position   = USI_Variable_Solutions::$options['preferences']['menu-position'];

      if ((0 == $position) || ('null' == $position)) $position = null;

      $this->page_hook = add_menu_page(
         __(USI_Variable_Solutions::NAME, USI_Variable_Solutions::TEXTDOMAIN), // Text displayed in <title> when menu is selected;
         __('Variables', USI_Variable_Solutions::TEXTDOMAIN), // Text displayed in menu; 
         $capability, // Capability required to enable page; 
         USI_Variable_Solutions::VARYLIST, // Unique slug to of this menu; 
         array($this, 'render_page'), // Function called to render page content;
         USI_Variable_Solutions::$options['preferences']['menu-icon'], // Menu icon;
         $position // Menu position;
      );

      add_action('load-' . $this->page_hook, array($this, 'action_load_screen_options'));
   
   } // action_admin_menu();

   function action_load_screen_options() {

      $option = 'per_page';

      $args = array(
         'label' => __('Variables per page', USI_Variable_Solutions::TEXTDOMAIN),
         'default' => 20,
         'option' => $option,
      );

      add_screen_option($option, $args);

      parent::__construct( 
         array(
            'singular' => __('variable', USI_Variable_Solutions::TEXTDOMAIN), 
            'plural' => __('variables', USI_Variable_Solutions::TEXTDOMAIN),
            'ajax' => false,
         ) 
      );

   } // action_load_screen_options();

   function column_cb($item) {

      return('<input class="' . USI_Variable_Solutions::VARYLIST . '"' .
         ' data-id="' . esc_attr($item['variable_id']) . '"' .
         ' data-name="' . esc_attr($item['variable']) . '"' .
         ' data-value="' . esc_attr($item['value']) . '"' .
         ' name="variable_id[]" type="checkbox" value="' . $item['variable_id'] .'" />');

    } // column_cb();

   function column_default($item, $column_name) {

      switch($column_name) { 
      case 'variable_id':
      case 'category':
      case 'notes':
      case 'owner':
      case 'type':
      case 'value':
      case 'variable':
         return $item[$column_name];
      case 'updated':
         $updated = strtotime($item['updated']);
         if ((abs(time() - $updated)) < 86400){
            return(sprintf('%s ago', human_time_diff($updated, strtotime(current_time('mysql')))));
         } else {
            return(date_i18n('j F, Y \a\t G:i', $updated));
         }
      default:
         return(print_r($item, true)); //Show the whole array for troubleshooting purposes
      }

   } // column_default();

   function column_variable($item) {

      $actions = array();

      if (USI_Variable_Solutions::$variables_change || USI_Variable_Solutions::$variables_edit) {
         $actions['edit'] = '<a href="options-general.php?page=usi-variable&variable_id=' .
            $item['variable_id'] . '">' . __('Edit', USI_Variable_Solutions::TEXTDOMAIN) . '</a>';
      }

      if (USI_Variable_Solutions::$variables_delete) {
         $actions['delete'] = '<a' .
            ' class="thickbox usi-variable-delete-link"' .
            ' data-id="' . esc_attr($item['variable_id']) . '"' .
            ' data-name="' . esc_attr($item['variable']) . '"' .
            ' data-value="' . esc_attr($item['value']) . '"' .
            ' href=""' .
            '">' . __('Delete', USI_Variable_Solutions::TEXTDOMAIN) . '</a>';
      }

      return($item['variable'] . ' ' . $this->row_actions($actions));

   } // column_variable();

   function extra_tablenav($which) {

      global $wpdb;

      if ('top' == $which) {        
         $SAFE_variables_table = $wpdb->prefix . 'USI_variables';
         $rows = $wpdb->get_results("SELECT DISTINCT `category` FROM `$SAFE_variables_table` WHERE (`category` <> '') ORDER BY `category`", OBJECT_K);
         echo '      <div class="alignleft actions bulkactions"><select id="' . $this->options_category . 
            '" name="' . $this->options_category . '">' .
            '<option ' . (($this->all_categories == $this->category) ? 'selected="selected" ' : '') . 'value="' . 
            $this->all_categories . '">' . $this->all_categories . '</option>';
         foreach ($rows as $row) {
            echo '<option ' . (($row->category == $this->category) ? 'selected="selected" ' : '') . 'value="' . $row->category . '">' . $row->category . '</option>';
         }
         echo '</select><input class="button action" type="submit" name="usi-filter" id="usi-filter" value="Filter" /></div>' . PHP_EOL;
      }

   } // extra_tablenav();

   function filter_set_screen_options($status, $option, $value) {

      if ('per_page' == $option) return($value);

      return($status);

   } // filter_set_screen_options();

   function get_bulk_actions() {

      return(
         array(
            'delete' => __('Delete', USI_Variable_Solutions::TEXTDOMAIN),
         )
      );

   } // get_bulk_actions();

   function get_columns() {

      return(
         array(
            'cb' => '<input type="checkbox" />',
            'variable_id' => __('ID', USI_Variable_Solutions::TEXTDOMAIN),
            'variable' => __('Variable', USI_Variable_Solutions::TEXTDOMAIN),
            'category' => __('Category', USI_Variable_Solutions::TEXTDOMAIN),
            'type' => __('Type', USI_Variable_Solutions::TEXTDOMAIN),
            'value' => __( 'Value', USI_Variable_Solutions::TEXTDOMAIN),
            'notes' => __('Notes', USI_Variable_Solutions::TEXTDOMAIN),
            'owner' => __('Owner', USI_Variable_Solutions::TEXTDOMAIN),
            'updated' => __('Updated', USI_Variable_Solutions::TEXTDOMAIN),
         )
      );

    } // get_columns();

   public function get_hidden_columns() {

      return((array)get_user_option('manage' . $this->page_hook . 'columnshidden'));

   } // get_hidden_columns();

   function get_list() {

      global $wpdb;

      $paged = (int)(isset($_GET['paged']) ? $_GET['paged'] : 1);

      $filter = $this->safe_name(isset($_POST['usi-filter']) ? $_POST['usi-filter'] : '');
      if ('Filter' == $filter) $paged = 1;

      $SAFE_order = (isset($_GET['order'])) ? (('desc' == strtolower($_GET['order'])) ? 'DESC' : '') : '';
      $WILD_orderby = (isset($_GET['orderby']) ? $_GET['orderby'] : '');
      switch ($WILD_orderby) {
      default: $SAFE_orderby = 'category` ' . $SAFE_order . ', `variable'; break;
      case 'notes': 
      case 'owner': 
      case 'type':
      case 'updated':
      case 'variable':
      case 'variable_id': $SAFE_orderby = $WILD_orderby;
      }

      $SAFE_orderby = 'ORDER BY `' . $SAFE_orderby . '` ' . $SAFE_order;
      $SAFE_search = ((isset($_POST['s']) && ('' != $_POST['s'])) ? $wpdb->prepare(' AND (`variable` = %s)', $_POST['s']) : '');
      if ('' == $SAFE_search) {
         if (!empty($_POST[$this->options_category])) {
            $this->category = $_POST[$this->options_category];
         } else if (!empty($_GET['filter'])) {
            $this->category = $_GET['filter'];
         } else {
            $category = get_user_option($this->options_category);
            $this->category = (!empty($category) ? $category : $this->all_categories);
         }
         update_user_option(get_current_user_id(), $this->options_category, $this->category);
         if ($this->all_categories != $this->category) $SAFE_search = $wpdb->prepare(' AND (`category` = %s)', $this->category);
      }

      $current_page = $this->get_pagenum();
      $SAFE_per_page = (int)$this->get_items_per_page('per_page', 20);
      $SAFE_skip = (int)($SAFE_per_page * ($paged - 1));

      $SAFE_variables_table = $wpdb->prefix . 'USI_variables';
      $count_of_records = $wpdb->get_var("SELECT COUNT(*) FROM `$SAFE_variables_table` WHERE (`variable_id` > 1)$SAFE_search");

      $SAFE_users_table = $wpdb->prefix . 'users';
      $this->items = $wpdb->get_results(
         "SELECT `variable_id`, `category`, `variable`, `type`, `value`, `display_name` as `owner`, " .
         "`$SAFE_variables_table`.`updated`, `$SAFE_variables_table`.`notes` FROM `$SAFE_variables_table`" .
         " INNER JOIN `$SAFE_users_table` ON `$SAFE_users_table`.`ID` = `$SAFE_variables_table`.`user_id`" . 
         " WHERE (`variable_id` > 1)$SAFE_search $SAFE_orderby LIMIT $SAFE_skip,$SAFE_per_page", ARRAY_A);

      for ($ith = 0; $ith < count($this->items); $ith++) {
         $this->items[$ith]['type'] = ('V' == $this->items[$ith]['type'] ? 'Variable' : 'Expression');
         $this->items[$ith]['value'] = htmlentities($this->items[$ith]['value']);
      }

      $this->set_pagination_args(
         array(
            'total_items' => $count_of_records,
            'per_page' => $SAFE_per_page,
            'total_pages' => ceil($count_of_records / $SAFE_per_page),
         )
      );

   } // get_list();

   function get_sortable_columns() {

      return(
         array(
            'variable_id' => array('variable_id', true),
            'category' => array('category', true),
            'variable' => array('variable', false),
            'notes' => array('notes', false),
            'owner' => array('owner', false),
            'type' => array('type', false),
            'updated' => array('updated', false),
         )
      );

   } // get_sortable_columns();

   function no_items() {

      _e('No variables have been configured.', USI_Variable_Solutions::TEXTDOMAIN);

    } // no_items();

   function prepare_items() {

      $columns = $this->get_columns();
      $hidden = $this->get_hidden_columns();
      $sortable = $this->get_sortable_columns();
      $this->_column_headers = array($columns, $hidden, $sortable);
      $this->get_list();

   } //prepare_items():

   function render_page() {

      global $wpdb;

      $action = $this->current_action();

      if ('delete' == $action) {
         if (USI_Variable_Solutions::$variables_delete) {
            $SAFE_variable_table = $wpdb->prefix . 'USI_variables';
            $ids = isset($_REQUEST['variable_id']) ? explode(',', $_REQUEST['variable_id']) : array();
            $variables_deleted = count($ids);
            if (is_array($ids)) $ids = implode(',', $ids);
            if (!empty($ids)) {
               $wpdb->query("DELETE FROM `$SAFE_variable_table` WHERE (`variable_id` IN($ids))");
            } else {
               $variables_deleted = 0;
            }
            $delete_text = ((1 == $variables_deleted) ? __('One variable has been deleted', USI_Variable_Solutions::TEXTDOMAIN) : 
               sprintf(__('%d variables have been deleted', USI_Variable_Solutions::TEXTDOMAIN), $variables_deleted));
         } else {
            $delete_text =  __('You do not have permission to delete variables', USI_Variable_Solutions::TEXTDOMAIN);
         }
         $message = '<div class="updated below-h2 notice is-dismissible" id="message"><p>' . $delete_text . '.</p>' .
            '<button type="button" class="notice-dismiss"><span class="screen-reader-text">' .
            __('Dismiss this notice', USI_Variable_Solutions::TEXTDOMAIN) . '.</span></button></div>';
      } else {
         $message = null;
      }
?>

<!-- usi-variable-solutions:render_page:begin ---------------------------------------------------------------------------------- -->
<div class="wrap">
  <h2><?php 
   _e('Variables', USI_Variable_Solutions::TEXTDOMAIN); 
   if (USI_Variable_Solutions::$variables_add) 
      echo ' <a class="add-new-h2" href="options-general.php?page=usi-variable">' . 
         __('Add New', USI_Variable_Solutions::TEXTDOMAIN) . '</a>';
   if (USI_Variable_Solutions::$variables_publish) 
      echo ' <a class="add-new-h2" href="options-general.php?page=usi-variable-settings&tab=publish">' . 
         __('Publish', USI_Variable_Solutions::TEXTDOMAIN) . '</a>';
  ?></h2>
  <?php if ($message) echo $message . PHP_EOL;?>
  <form action="" method="post" name="<?php echo USI_Variable_Solutions::VARYLIST;?>">
    <input type="hidden" name="page" value="<?php echo $_REQUEST['page'];?>">
<?php
      $this->prepare_items(); 
      $this->search_box('search', 'search_id');
      $this->display(); 
?>
  </form>
</div>
<div id="usi-variable-confirm" style="display:none;"></div>
<script>
jQuery(document).ready(
   function($) {

      var text_confirm_prefix = '<?php _e("Please confirm that you want to delete the following variable(s)", USI_Variable_Solutions::TEXTDOMAIN);?>';
      var text_confirm_suffix = '<?php _e("This deletion is permanent and cannot be reversed", USI_Variable_Solutions::TEXTDOMAIN);?>';
      var text_cancel = '<?php _e("Cancel", USI_Variable_Solutions::TEXTDOMAIN);?>';
      var text_delete = '<?php _e("Delete", USI_Variable_Solutions::TEXTDOMAIN);?>';
      var text_ok     = '<?php _e("Ok", USI_Variable_Solutions::TEXTDOMAIN);?>';
      var text_please_action   = '<?php _e("Please select a bulk action before you click the Apply button.", USI_Variable_Solutions::TEXTDOMAIN);?>';
      var text_please_variable = '<?php _e("Please select one or more variables before you click the Apply button.", USI_Variable_Solutions::TEXTDOMAIN);?>';

      function do_action() {

         var ids = $('.usi-variable');
         var id_list = '';
         var text = '';

         if ('delete' != $('#bulk-action-selector-top').val()) {
            text = text_please_action;
         } else {
            var delete_count = 0;
            for (var i = 0; i < ids.length; i++) {
               if (ids[i].checked) {
                  id_list += (id_list.length ? ',' : '') + ids[i].getAttribute('data-id');
                  text += (delete_count++ ? '<br/>' : '') + ids[i].getAttribute('data-name') + ' = ' + ids[i].getAttribute('data-value');
               }
            }
            if (!delete_count) {
               text = text_please_variable;
            }
         }

         return(show_confirmation(delete_count, id_list, text));

      } // do_action();

      function show_confirmation(count_of_variables, id, text) {

         var html = '<p>';

         if (count_of_variables) {
            html += text_confirm_prefix + ':</p><p>' + text + '</p><p>' + text_confirm_suffix + '.';
         } else {
            html += text;
         }

         html += '</p><hr/><p>';

         if (count_of_variables) html += 
            '<a class="button" href="?page=<?php echo USI_Variable_Solutions::VARYLIST;?>&action=delete&variable_id=' +
            id + '">' + text_delete + '</a> &nbsp; ';

         html += '<a class="button" href="" onclick="tb_remove()">' +
            (count_of_variables ? text_cancel : text_ok) + '</a>';

         $('#usi-variable-confirm').html(html);

         tb_show('Variable-Solutions', '#TB_inline?width=500&height=300&inlineId=usi-variable-confirm', null);

         return(false);

      } // show_confirmation();

      $('#doaction').click(do_action); 

      $('#doaction2').click(do_action); 

      $('.usi-variable-delete-link').click(
         function(event) {
            var obj = event.target;
            var id = obj.getAttribute('data-id');
            var text = obj.getAttribute('data-name') + ' = ' + obj.getAttribute('data-value');
            return(show_confirmation(1, id, text));
         }
      );

   }
);
</script>
<!-- usi-variable-solutions:render_page:end ------------------------------------------------------------------------------------ -->
<?php
   } // render_page();

   static function safe_name($category) {

      return(strtolower(str_replace('-', '_', sanitize_title($category))));

   } // safe_name();

} // Class USI_Variable_Solutions_Table;

new USI_Variable_Solutions_Table();

// --------------------------------------------------------------------------------------------------------------------------- // ?>
